RecoMuon/L2MuonSeedGenerator/src/Phase2L2MuonSeedCreator.cc

```diff
@@ -0,0 +1,842 @@
+      extrapolationDeltaPhiFar_{pset.getParameter<double>("extrapolationWindowFar")},
+      maxEtaBarrel_{pset.getParameter<double>("maximumEtaBarrel")},
+      maxEtaOverlap_{pset.getParameter<double>("maximumEtaOverlap")},
+      propagatorName_{pset.getParameter<string>("Propagator")} {
+  // Service parameters
+  edm::ParameterSet serviceParameters = pset.getParameter<edm::ParameterSet>("ServiceParameters");
+  // Services
+  service_ = std::make_unique<MuonServiceProxy>(serviceParameters, consumesCollector());
+  estimator_ = std::make_unique<Chi2MeasurementEstimator>(10000.);
+  produces<L2MuonTrajectorySeedCollection>();
+}
+
+void Phase2L2MuonSeedCreator::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("InputObjects", edm::InputTag("l1tTkMuonsGmt"));
+  desc.add<edm::InputTag>("CSCRecSegmentLabel", edm::InputTag("hltCscSegments"));
+  desc.add<edm::InputTag>("DTRecSegmentLabel", edm::InputTag("hltDt4DSegments"));
+  desc.add<double>("MinPL1Tk", 3.5);
+  desc.add<double>("MaxPL1Tk", 200);
+  desc.add<double>("stubMatchDPhi", 0.05);
+  desc.add<double>("stubMatchDTheta", 0.1);
+  desc.add<double>("extrapolationWindowClose", 0.1);
+  desc.add<double>("extrapolationWindowFar", 0.05);
+  desc.add<double>("maximumEtaBarrel", 0.7);
+  desc.add<double>("maximumEtaOverlap", 1.3);
+  desc.add<string>("Propagator", "SteppingHelixPropagatorAny");
+
+  // Service parameters
+  edm::ParameterSetDescription psd0;
+  psd0.addUntracked<std::vector<std::string>>("Propagators", {"SteppingHelixPropagatorAny"});
+  psd0.add<bool>("RPCLayers", true);
+  psd0.addUntracked<bool>("UseMuonNavigation", true);
+  desc.add<edm::ParameterSetDescription>("ServiceParameters", psd0);
+  descriptions.add("Phase2L2MuonSeedCreator", desc);
+}
+
+void Phase2L2MuonSeedCreator::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  const std::string metname = "RecoMuon|Phase2L2MuonSeedCreator";
+
+  auto output = std::make_unique<L2MuonTrajectorySeedCollection>();
+
+  auto const l1TkMuColl = iEvent.getHandle(l1TkMuCollToken_);
+
+  auto cscHandle = iEvent.getHandle(cscSegmentCollToken_);
+  auto cscSegments = *cscHandle;
+  auto dtHandle = iEvent.getHandle(dtSegmentCollToken_);
+  auto dtSegments = *dtHandle;
+
+  cscGeometry_ = iSetup.getHandle(cscGeometryToken_);
+  dtGeometry_ = iSetup.getHandle(dtGeometryToken_);
+  magneticField_ = iSetup.getHandle(magneticFieldToken_);
+
+  LogDebug(metname) << "Number of L1 Tracker Muons in the Event: " << l1TkMuColl->size();
+
+  // Loop on all L1TkMu in event
+  for (size_t l1TkMuIndex = 0; l1TkMuIndex != l1TkMuColl->size(); ++l1TkMuIndex) {
+    l1t::TrackerMuonRef l1TkMuRef(l1TkMuColl, l1TkMuIndex);
+
+    // Physical info of L1TkMu
+    float pt = l1TkMuRef->phPt();
+    if (pt < minMomentum_) {
+      continue;
+    }
+    float eta = l1TkMuRef->phEta();
+    float theta = 2 * std::atan(std::exp(-eta));
```

Comment:
this pattern together with the usage of `ThreeVector` from `CLHEP` smells.
What about something possibly more efficient reducing the usage of trig functions along the lines of https://github.com/cms-sw/cmssw/commit/6272d52863ba88908a84ad05ab1de072450a29d0 ?

Reply:
The latest commit should address this. It's still computing theta (only once per L1TkMu) since it's used for the stub-segment matching, but everything has been refactored to result in fewer overall "taxing" computations